R/SandwichLayerVariance.R

```diff
@@ -558,3 +559,459 @@ vcov_tee <- function(x, type = "CR0", cluster = NULL, ...) {
+#' @details Calculate upper and lower bound estimates of meat matrix 
+#' for design-based variance estimate for \code{teeMod} models 
+#' with covariance adjustment and without absorbed effects
+#' @return a list of meat matrix bounds
+#' @keywords internal
+.get_DB_covadj_meat <- function(x, ...) {
+  agg <- .aggregate_to_cluster(x)
+  data <- agg[[1]]
+  bid <- data[, agg[[2]]]  # block ids
+  zobs <- data[, agg[[3]]] # observed zs
+  
+  p <- length(x$call$offset@fitted_covariance_model$coefficients)
+  XX <- .prepare_design_matrix(x)
+  
+  V00 <- .cov_mat_est(XX[zobs==0,], bid[zobs==0])
+  V11 <- .cov_mat_est(XX[zobs==1,], bid[zobs==1])
+  V01 <- .cov01_est(XX, zobs, bid)
+  
+  idl <- (p+2):(2*p+1)
+  meat1u <- V00[1:p, 1:p] + V01[1:p, 1:p] + t(V01[1:p, 1:p]) + V11[1:p, 1:p]
+  meat1l <- V00[idl, 1:p] + V01[idl, 1:p] + t(V01[idl, 1:p]) + V11[idl, 1:p]
+  
+  meat2u <- cbind(V00[1:p, p+1], V01[1:p, p+1]) + cbind(V01[p+1, 1:p], V11[1:p, p+1])
+  meat2l <- cbind(V00[idl, p+1], V01[idl, p+1]) + cbind(V01[2*p+2, 1:p], V11[idl, p+1])
+  
+  meat3u <- matrix(c(V00[p+1, p+1], V01[p+1, p+1], V01[p+1, p+1], V11[p+1, p+1]), ncol = 2)
+  meat3l <- matrix(c(V00[2*p+2, p+1], V01[2*p+2, p+1], V01[2*p+2, p+1], V11[2*p+2, p+1]), ncol = 2)
+  
+  return(list(m1u = meat1u, m1l = meat1l,
+              m2u = meat2u, m2l = meat2l,
+              m3u = meat3u, m3l = meat3l))
+}
+
+#' @title (Internal) Design-based variance for models without covariance adjustment
+#' @param x a fitted \code{teeMod} model
+#' @details Calculate bread matrix for design-based variance estimate for 
+#'  \code{teeMod} models without covariance adjustment and 
+#'  without absorbed effects
+#' @return design-based variance estimate of the main treatment effect estimate
+#' @keywords internal
+.get_DB_wo_covadj_se <- function(x, ...){
+  if (x@absorbed_intercepts) {
+    stop("x should not have absorbed intercepts")
+  }
+  if (!is.null(x$call$offset)){
+    stop("x should not have covariance adjustment")
+  }
+  agg <- .aggregate_to_cluster(x)
+  data <- agg[[1]]
+  block <- agg[[2]]
+  
+  ws <- data$.w
+  yobs <- data$.wy  # observed ys by cluster
+  zobs <- data[, agg[[3]]]  # observed zs by cluster
+  bid <- data[, block] # block ids of clusters
+  
+  nbk <- design_table(design=x@Design, x="treatment",y="block")
+  # number of units by block and treatment, B by K
+  small_blocks <- apply(nbk, 1, min) == 1 # small block indicator
+  
+  nb <- rowSums(nbk)
+  rho <- c(sum(ws * yobs * (1-zobs)) / sum(ws * (1-zobs)),
+           sum(ws * yobs * zobs) / sum(ws * zobs))
+  
+  gammas <- (nbk[bid,1]*(1-zobs) + nbk[bid,2]*zobs) * ws # pseudo outcomes
+  gamsbk <- list()  # s^2_b,j, sample variance
+  for (k in 1:2){
+    indk <- zobs == (k-1)
+    gammas[indk] <- gammas[indk] * (yobs[indk] - rho[k])
+    gamsbk[[k]] <- aggregate(gammas[indk], by = list(data[indk,block]), FUN = var)
+  }
+  gamsbk <- merge(gamsbk[[1]], gamsbk[[2]], by = "Group.1")[,2:3]
```

Comment:
Same note as below about carefully testing functions that use `merge()`

Reply:
Added a few more test cases (commit https://github.com/benbhansen-stats/propertee/commit/7fbf8e2e103f0d7191d96ceb45987d08318f1aed)